test/server/server_test.cc

```diff
@@ -197,10 +237,74 @@ class ServerInstanceImplTest : public testing::TestWithParam<Network::Address::I
   std::unique_ptr<InstanceImpl> server_;
 };
 
+// Custom StatsSink that just increments a counter when flush is called.
+class CustomStatsSink : public Stats::Sink {
+public:
+  CustomStatsSink(Stats::Scope& scope) : stats_flushed_(scope.counter("stats.flushed")) {}
+
+  // Stats::Sink
+  void flush(Stats::MetricSnapshot&) override { stats_flushed_.inc(); }
+
+  void onHistogramComplete(const Stats::Histogram&, uint64_t) override {}
+
+private:
+  Stats::Counter& stats_flushed_;
+};
+
+// Custom StatsSinFactory that creates CustomStatsSink.
+class CustomStatsSinkFactory : public Server::Configuration::StatsSinkFactory {
+public:
+  // StatsSinkFactory
+  Stats::SinkPtr createStatsSink(const Protobuf::Message&, Server::Instance& server) override {
+    return std::make_unique<CustomStatsSink>(server.stats());
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{new Envoy::ProtobufWkt::Empty()};
+  }
+
+  std::string name() override { return "envoy.custom_stats_sink"; }
+};
+
 INSTANTIATE_TEST_SUITE_P(IpVersions, ServerInstanceImplTest,
                          testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
                          TestUtility::ipTestParamsToString);
 
+/**
+ * Static registration for the custom sink factory. @see RegisterFactory.
+ */
+REGISTER_FACTORY(CustomStatsSinkFactory, Server::Configuration::StatsSinkFactory);
+
+// Validates that server stats are flushed even when server is stuck with initialization.
+TEST_P(ServerInstanceImplTest, StatsFlushWhenServerIsStillInitializing) {
+  absl::Notification started;
+
+  auto server_thread = Thread::threadFactoryForTest().createThread([&] {
+    initialize("test/server/stats_sink_bootstrap.yaml", true);
+    auto startup_handle = server_->registerCallback(ServerLifecycleNotifier::Stage::Startup,
+                                                    [&] { started.Notify(); });
+    auto shutdown_handle = server_->registerCallback(ServerLifecycleNotifier::Stage::ShutdownExit,
+                                                     [&](Event::PostCb) { FAIL(); });
+    shutdown_handle = nullptr; // unregister callback
+    server_->run();
+    startup_handle = nullptr;
+    server_ = nullptr;
+    thread_local_ = nullptr;
+  });
+
+  started.WaitForNotification();
+
+  // Wait till stats are flushed to custom sink and validate its value.
+  while (TestUtility::findCounter(stats_store_, "stats.flushed") == nullptr ||
```

Comment:
Can you potentially just use `waitForCounterEq` or factor it in a way that it can be used here?

Reply:
I thought about that. We can not use that directly but we can refactor. I was actually thinking of doing a follow-up PR to move all `waitForxxxx` methods to TestUtility so that they can be reused better. For now I will refactor watiForCounter and do a follow-up for others